py/gc.c

```diff
@@ -796,6 +818,29 @@ void *gc_realloc(void *ptr_in, size_t n_bytes, bool allow_move) {
 }
 #endif // Alternative gc_realloc impl
 
+#if MICROPY_GC_PERMANENT_ALLOCS
+void gc_set_permanent(void *ptr, bool permanent) {
+    // Add/remove from list.
+    if (permanent) {
+        if (MP_STATE_MEM(gc_permanent_used) == MP_STATE_MEM(gc_permanent_alloc)) {
+            MP_STATE_MEM(gc_permanent_alloc) += WORDS_PER_BLOCK;
+            MP_STATE_MEM(gc_permanent) = gc_realloc(MP_STATE_MEM(gc_permanent), MP_STATE_MEM(gc_permanent_alloc) * sizeof(void*), true);
+        }
+        MP_STATE_MEM(gc_permanent)[MP_STATE_MEM(gc_permanent_used)++] = ptr;
+    } else {
+        for (size_t i = 0; i < MP_STATE_MEM(gc_permanent_used);) {
+            if (MP_STATE_MEM(gc_permanent)[i] == ptr) {
+                // Shift back one element.
+                --MP_STATE_MEM(gc_permanent_used);
+                memmove(MP_STATE_MEM(gc_permanent) + i, MP_STATE_MEM(gc_permanent) + i + 1, sizeof(void*) * (MP_STATE_MEM(gc_permanent_used) - i));
```

Comment:
break after this?

Reply:
Leaving the break out means that it handles duplicates. Which shouldn't happen, but the "set" path doesn't prevent it.

Reply:
In that case, instead of using memmove, it could shift the values down as it does the search (ie implement memmove as part of the search).  But that only makes sense if the implementation of that can be minimal, which it might be.

Reply:
Done. It saves 8 bytes on Cortex M4.